sdk/file_sdk/XMPFileFormat/xmpio_over_istream.cpp

```diff
@@ -0,0 +1,85 @@
+#include "xmpio_over_istream.h"
+
+namespace mip {
+namespace file {
+
+XMPIOOverIStream::XMPIOOverIStream(std::shared_ptr<IStream> stream) {
+  mBaseStream = stream;
+}
+
+XMPIOOverIStream::~XMPIOOverIStream() {
+  if (mTemp != nullptr)
+    delete mTemp;
+}
+
+XMP_Uns32 XMPIOOverIStream::Read(void* buffer, XMP_Uns32 count, bool readAll) {
+  if (buffer == nullptr)
+    throw new XMP_Error(kXMPErr_BadParam, "Buffer is null");
+
+  auto readCount = mBaseStream->Read(static_cast<uint8_t*>(buffer), count);
+  if(readAll && (XMP_Uns32)readCount < count)
+    throw new XMP_Error(kXMPErr_BadParam, "readCount is lower than expected");

```

Comment:
>kXMPErr_BadParam [](start = 24, length = 16)

is this badparam error or something like kXMPErr_ReadError?

Reply:
does it matter? i think kXMPErr_BadParam is more suitable 

---
In reply to: [137464437](https://github.com/AzureAD/rms-sdk-for-cpp/pull/140#discussion_r137464437) [](ancestors = 137464437)

Reply:
bad param is when the parameter is wrong. this is not the case.
see: [https://github.com/01org/umf/blob/dff137b2f9e1c283f9a3c2af4b59212edda10fcd/3rdparty/xmp/source/XMPFiles_IO.cpp#L144](https://github.com/01org/umf/blob/dff137b2f9e1c283f9a3c2af4b59212edda10fcd/3rdparty/xmp/source/XMPFiles_IO.cpp#L144)
they are using kXMPErr_EnforceFailure 

---
In reply to: [137476388](https://github.com/AzureAD/rms-sdk-for-cpp/pull/140#discussion_r137476388) [](ancestors = 137476388,137464437)

Reply:
The code in this link is not official, every one can implement it as they want but if you think kXMPErr_EnforceFailure is better, i will change it

---
In reply to: [137487775](https://github.com/AzureAD/rms-sdk-for-cpp/pull/140#discussion_r137487775) [](ancestors = 137487775,137476388,137464437)